### PR TITLE
localStorage usage Guideline

### PR DESCRIPTION
**Pull request by zhouyx**
There were several discussions on this topic, hope this clarifies our policy on future localStorage usage. Let me know if this is a good place to add the doc. 

cc @cramforce @mrjoro

### CODE REVIEW

File: spec/amp-localstorage.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+- Client ID Service : Store the user decision to opt-out CID service.
+
+## Guidance on localStorage Usage in AMP
+
+The localStorage can be used by future AMP components or services with valid reaons to store info client side.
+
+There are a few guidlines to follow due to localStorage implementation in AMP and AMP Viewer.
+
+### Privacy Policy
+Entries can be sent to the AMP Viewer and stored in the top level window. No peronsal identifiable information would be allowed to comply with our privacy policy. We require that the storage entry to be structured with source code checked into the open source library. Neither the storage key or the value can be calculated externally.
```
Comment (riklund) [inline spec/amp-localstorage.md]:
```suggestion
Entries can be sent to the AMP Viewer and stored in the top level window. No personal identifiable information would be allowed to comply with our privacy policy. We require that the storage entry to be structured with source code checked into the open source library. Neither the storage key or the value can be calculated externally.
```

**Comment by jeffjose** [comment]
Mixed usage of `Viewer`, `AMP Viewer` and `AMP viewer`. Please pick one and keep it consistent.

File: spec/amp-localstorage.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS-IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# LocalStorage Usage in AMP
+
+AMP relies on the Viewer's support to use localStorage when the AMP page is embedded in a Viewer. This is because localStorage set within a cross domain iframe is not persisted in Safari.
```
Comment (lannka) [inline spec/amp-localstorage.md]: safari => some browsers

File: spec/amp-localstorage.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+blob = btoa(JSON.stringify(entry))
+```
+
+## Current Usage
+
+The following AMP components and service are using the localStorage.
+
+- `<amp-app-banner>` : Store the user decision on dismiss banner
+- `<amp-user-notification>` : Store the user decision on dismiss notification
+- `<amp-consent>` : Store the user decision and granular information on consent
```
Comment (lannka) [inline spec/amp-localstorage.md]: "any future usage need to be listed here"

Comment (zhouyx) [reply spec/amp-localstorage.md]: That's a good point. But really hard to keep the doc up to date though. 
Best I can do is to add one sentence to the last session.

File: spec/amp-localstorage.md (hunk 1 above)
Comment (zhouyx) [reply spec/amp-localstorage.md]: nice catch 😄 fixed

File: spec/amp-localstorage.md (hunk 2 above)
Comment (zhouyx) [reply spec/amp-localstorage.md]: done

File: spec/amp-localstorage.md (hunk 3 above)
Comment (zhouyx) [reply spec/amp-localstorage.md]: Please see if the new change looks good to you. thanks

**Comment by zhouyx** [comment]
Thank you for the review. @jeffjose switched to AMP Viewer.

@lannka Added one paragraph on falling back to `window.localStorage`. ptal. Thanks

File: spec/amp-localstorage.md (hunk 3 above)
Comment (lannka) [reply spec/amp-localstorage.md]: we can add a link in the presubmit warning:
https://github.com/ampproject/amphtml/blob/75e7d72be80d9bd356ad76e0a419cd2539320e6b/build-system/tasks/presubmit-checks.js#L27-L30

Comment (zhouyx) [reply spec/amp-localstorage.md]: 👍 That's a great idea. Done, ptal

**Comment by zhouyx** [comment]
Thank you all for reviewing : )

**Pull request merged**